Add ViewPager to Entry view

SEAbdulbasit opened the pull request:
## :scroll: Description
<!-- Describe your changes in detail, add links to issue/ticket -->
I have added another Fragment for ViewPager. First, we retrieve all the entries and submit them to ViewPager. Then we find the selected date and move to that page position. 
Converted `parentFragmentManager` to `supportFragmentManager` I had difficulty popping up the view pager fragment.

Ticket: https://github.com/alisonthemonster/Presently/issues/161
## :green_heart: How did you test it?
<!-- Unit, Integration, Manual? If manual, how was it tested? -->
- Unit test it 
- Did some manual testing. Added new entries from Timeline Screen and also from Calendar. Tested the share functionality along with search. 

## :camera_flash: Screenshots / GIFs
<!-- Mandatory for UI changes -->
<!-- Uncomment the next line and replace it with links to your screenshots. -->

https://user-images.githubusercontent.com/33172684/188437796-9e31e05c-6489-44c4-ad06-d8ef5960e3e0.mp4


<!-- <img src="https://user-images.githubusercontent.com/883468/93824506-8ef15880-fc31-11ea-9f44-ca32bb0de88a.jpg" width="260"/> -->

SEAbdulbasit committed:
View Pager added to Entry View

SEAbdulbasit committed:
Unit test added for EntryViewPagerViewModel.kt
Directory name changed

SEAbdulbasit commented:
While I was creating the video, another use case popped up in my head, what if we edit and then swipe ?
As data will be lost, Is it okay or should we also should a dialogue on swipe incase of edit ?

alisonthemonster commented on app/src/main/java/journal/gratitude/com/gratitudejournal/ui/entryviewpager/ViewPagerAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+
+        return EntryFragment.newInstance(
+            date = itemsList[position].entryDate,
+            numEntries = 0,
```
I think we want to use the `numEntries` variable you created instead of zero here! Otherwise the milestone dialog won't appear when the user saves a milestone creating entry!

alisonthemonster commented on app/src/main/java/journal/gratitude/com/gratitudejournal/ui/entryviewpager/EntryViewPagerViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+                            itemsList.add(1, Entry(yesterday, ""))
+                        }
+                    }
+                }
```
This code is duplicated from `TimelineViewModel` maybe we can move it into a util somewhere?

alisonthemonster commented on app/src/main/java/journal/gratitude/com/gratitudejournal/ui/entryviewpager/EntryViewPagerFragment.kt:
```diff
@@ -0,0 +1,77 @@
+        super.onCreate(savedInstanceState)
+        val callback = object : OnBackPressedCallback(true) {
+            override fun handleOnBackPressed() {
+                withState(viewModel) {
```
Do we need the state here?

alisonthemonster commented on app/src/main/java/journal/gratitude/com/gratitudejournal/ui/entryviewpager/ViewPagerAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+        val item = itemsList[position]
+        var numEntries = 0
+        for (entry in itemsList) {
+            if (entry is Entry && entry.entryContent.isNotEmpty()) numEntries++
```
I don't think we need the `entry is Entry` check here!

alisonthemonster commented:
Thank you so much for adding this feature!! Apologies I didn't get to review it sooner, I've kicked off the CI builds and I'm reviewing now :) 

> While I was creating the video, another use case popped up in my head, what if we edit and then swipe ? As data will be lost, Is it okay or should we also should a dialogue on swipe incase of edit ?

That's a great point! I think a dialog in the case where they've typed and swipe away sounds perfect!

alisonthemonster reviewed (commented):
CI is failing because we've dropped below code coverage requirements. I think [this link](https://output.circle-artifacts.com/output/job/1d0a368f-caa4-40e2-91c1-af94d585e7fe/artifacts/0/results/jacoco/html/index.html) will work, it shows the current code coverage for the app.

If we could add some instrumented tests for the view pager UI then I think it would pass! Mavericks does create some default methods that jacoco doesn't understand aren't testable (I've opened up [this issue](https://github.com/airbnb/mavericks/issues/552) about it), but we'll see how close we can get!

SEAbdulbasit commented:
Regarding this https://github.com/alisonthemonster/Presently/pull/261#issuecomment-1241430757 Showing the dialog,I was looking into some solutions.
One solution might be to implement an interface at EntryView and then get the current fragment at the view pager and check if any edits were made when swiping. If edits were made, then we will cancel the swiping and show a dialogue. We cannot check at EnteryView because when swiped, the page will be changed and dialogue will be showing at the previous or next entry.
Would love to hear your thoughts on this solution.

SEAbdulbasit committed:
PR comments addressed
Dialog added incase of edits

SEAbdulbasit committed:
Entry List refactored

SEAbdulbasit committed:
Code refactoring

SEAbdulbasit committed:
Instrumental test added for EntryViewPagerFragment

SEAbdulbasit committed:
Variable name updated

alisonthemonster commented on app/src/main/java/journal/gratitude/com/gratitudejournal/ui/entry/EntryFragment.kt:
```diff
@@ -292,4 +300,32 @@ class EntryFragment : Fragment(), MavericksView {
+                return@withState true
+            }
+            return@withState false
+        }
```
Let's move this logic into the `EntryState` as a [derived property](https://airbnb.io/mavericks/#/core-concepts?id=derived) so you can just do `it.editsWereMade` or something like that and we can test this logic!

alisonthemonster commented on app/src/main/java/journal/gratitude/com/gratitudejournal/ui/entry/EntryFragment.kt:
```diff
@@ -292,4 +300,32 @@ class EntryFragment : Fragment(), MavericksView {
+
+}
+
+interface CanSwipe {
```
Can we rename this to some thing more like `EntryScreenCallbacks` or `EntryScreenActions`? Looking to get more clarity in the name.

alisonthemonster commented on strings/src/main/res/values/strings.xml:
```diff
@@ -72,6 +72,7 @@
     <string name="fingerprint_error_lockout_too_many">Too many biometric authentication attempts. Please try again in a few minutes.</string>
     <string name="continue_to_exit">Continue</string>
     <string name="are_you_sure">Are you sure you want to go back?</string>
+    <string name="are_you_sure_to_swipe">Are you sure you want to Swipe?</string>
```
Let's lowercase the word `Swipe`. We'll also need to add the translations for the other languages but I can take care of that after this PR is merged!

SEAbdulbasit replied:
If you want, I can translate it via Google Translator and add it for different languages, wdyt?

alisonthemonster replied:
If you don't mind going to all the trouble then please be my guest :)

alisonthemonster reviewed (commented):
Thank you for adding the warning dialog feature and for addressing my comments! I just ran on CI and there's a compile error in `EntryViewPagerViewModelTest`!

Other than that its looking good!

SEAbdulbasit committed:
Callback name update to EntryScreenCallbacks

SEAbdulbasit committed:
Entry ViewPager Test updated

SEAbdulbasit committed:
Edits check refactored to derived property

SEAbdulbasit committed:
Unit test added for derived property

SEAbdulbasit committed:
Swipe converted to small letters

SEAbdulbasit commented:
@alisonthemonster Thanks for the comments. I have addressed the comments and looked into the failing tests. I forgot to update the `EntryViewPagerViewModelTest` which is now done.

alisonthemonster commented:
We are so close to getting this landed! 76.55% is the current method coverage and we need 77! 

There are a few changes we can make to bump up coverage:
In `EntryViewPagerFragment` I think we can make `adapter` and `previousPosition` both private fields. In `ViewPagerAdapter` the var `numEntries` and the val `fragment` can be private too. 

But I think with using Mavericks and Jacoco we're just going to run into problems moving forward. I'm going to go ahead and get rid of the method coverage requirement. I think we're sufficiently tracking code coverage using line/instruction/class. I'll get that merged and then you can merge/rebase and we should be good to go!

SEAbdulbasit committed:
ViewPagerAdapter and EntryViewPager variables made private

SEAbdulbasit committed:
Converted == to isEmpty

SEAbdulbasit committed:
Merge branch 'alisonthemonster:develop' into feat/add-view-pager-to-entry-views

SEAbdulbasit commented:
@alisonthemonster you can run the coverage task. I have pulled the changes.

alisonthemonster commented:
Thank you for your patience! I'll get this merged and work on a release :)

SEAbdulbasit commented:
@alisonthemonster found a case. Incase of new entry, if swipe, will stuck in a loop as entry does not get reset.
Looking into it.

https://user-images.githubusercontent.com/33172684/196130140-f5d182ce-9d76-486d-9916-1ae856bb71a7.mp4
